fireant/queries/builder/dataset_blender_query_builder/klass.py

```diff
@@ -0,0 +1,387 @@
+        sub_query_tables = [dict() for _ in primary_tables]
+        sub_query_joins = [set() for _ in primary_tables]
+
+        for i, primary_table in enumerate(primary_tables):
+            sub_query_tables[i][self.primary_dataset.table] = primary_table
+            sub_query_tables[i][Table('blend_{}'.format(self.primary_dataset.table._table_name))] = primary_table
+
+            for secondary_dataset in self.secondary_datasets:
+                sub_query = queries_per_dataset[secondary_dataset.table].sql[i]
+                sub_query_tables[i][secondary_dataset.table] = sub_query
+                sub_query_tables[i][Table('blend_{}'.format(secondary_dataset.table._table_name))] = sub_query
+                sub_query_criterion, dimensions_without_mapping = self.__criterion_for_sub_query(
+                    primary_table, secondary_dataset, sub_query, self._dimensions,
+                )
+
+                sub_query_joins[i].add(Join(sub_query, join_type=JoinType.left, criterion=sub_query_criterion))
+
+        return sub_query_joins, sub_query_tables
+
+    @property
+    def __dimensions_per_dataset(self):
+        """
+        Returns a mapping between datasets and their respective dimensions.
+
+        :return: a dictionary with datasets as keys and their respective list of dimensions as values.
+        """
+        dimensions_per_dataset = defaultdict(list)
+
+        for dimension in self._dimensions:
+            for dataset in self.all_datasets:
+                new_dimension = copy.deepcopy(dimension)
+
+                if isinstance(dimension, DimensionModifier):
+                    new_dimension.dimension = self.dataset.field_mapping[dataset.table][
+                        new_dimension.dimension
+                    ]
+
+                new_dimension.definition = copy.deepcopy(new_dimension.definition)
+                for pypika_field in new_dimension.definition.fields():
+                    pypika_field.table = dataset.table
+                dimensions_per_dataset[dataset].append(new_dimension)
+
+        return dimensions_per_dataset
+
+    @property
+    def __metrics_per_dataset(self):
+        """
+        Returns a mapping between datasets and their respective metrics.
+
+        :return: a dictionary with datasets as keys and their respective list of metrics as values.
+        """
+        metrics_per_dataset = defaultdict(list)
+        metrics = find_metrics_for_widgets(self._widgets)
+
+        for metric in metrics:
+            for pypika_field in metric.definition.fields():
+                for dataset in self.all_datasets:
+                    try:
+                        metrics_per_dataset[dataset].append(
+                            dataset.fields[
+                                _get_pypika_field_name_or_alias(pypika_field).replace('_', '-').replace('$', '')
+                            ]
+                        )
+                    except AttributeError:
```

Comment:
could we put a comment to explain that these attributeerrors are meant to be ignored?

Reply:
Sure.